File: src/apps/agentic-support/services/metricsService.js
/**
 * Metrics service for fetching observability data
 * Author: Vinod Kumar V (VKV)
 */

const API_BASE_URL = "http://localhost:8000/api/v1/metrics";

/**
 * Format date to YYYY-MM-DD
 */
function formatDate(date) {
  if (!date) return null;
  if (typeof date === "string") return date;
  return date.toISOString().split("T")[0];
}

/**
 * Get date range for preset options
 */
function getDateRange(preset) {
  const now = new Date();
  const toDate = formatDate(now);
  
  switch (preset) {
    case "24h":
      const yesterday = new Date(now);
      yesterday.setHours(yesterday.getHours() - 24);
      return { from_date: formatDate(yesterday), to_date: toDate };
    case "7d":
      const weekAgo = new Date(now);
      weekAgo.setDate(weekAgo.getDate() - 7);
      return { from_date: formatDate(weekAgo), to_date: toDate };
    case "30d":
      const monthAgo = new Date(now);
      monthAgo.setDate(monthAgo.getDate() - 30);
      return { from_date: formatDate(monthAgo), to_date: toDate };
    case "90d":
      const quarterAgo = new Date(now);
      quarterAgo.setDate(quarterAgo.getDate() - 90);
      return { from_date: formatDate(quarterAgo), to_date: toDate };
    default:
      return { from_date: null, to_date: null };
  }
}

/**
 * Build query string from params
 */
function buildQueryString(params) {
  const query = new URLSearchParams();
  Object.entries(params).forEach(([key, value]) => {
    if (value !== null && value !== undefined) {
      query.append(key, value);
    }
  });
  const queryString = query.toString();
  return queryString ? `?${queryString}` : "";
}

/**
 * Fetch with error handling
 */
async function fetchMetrics(endpoint, params = {}) {
  try {
    const queryString = buildQueryString(params);
    const url = `${API_BASE_URL}${endpoint}${queryString}`;
    const response = await fetch(url);
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    return await response.json();
  } catch (error) {
    console.error(`Error fetching ${endpoint}:`, error);
    // Return mock data on error for demo purposes
    return getMockData(endpoint);
  }
}

/**
 * Get mock data for demo purposes
 */
function getMockData(endpoint) {
  if (endpoint === "/channels") {
    return {
      channels: [
        { channel: "voice", volume: 12847, auto_resolved: 8450, escalated: 2280 },
        { channel: "chat", volume: 9876, auto_resolved: 7200, escalated: 1200 },
        { channel: "email", volume: 4532, auto_resolved: 3200, escalated: 800 },
        { channel: "sms", volume: 2345, auto_resolved: 1800, escalated: 400 },
        { channel: "whatsapp", volume: 1234, auto_resolved: 900, escalated: 200 },
        { channel: "web", volume: 5678, auto_resolved: 4500, escalated: 800 },
        { channel: "mobile", volume: 3456, auto_resolved: 2800, escalated: 500 },
        { channel: "api", volume: 890, auto_resolved: 750, escalated: 100 },
      ],
    };
  }
  
  if (endpoint === "/kpi") {
    return {
      self_heal_rate: 78.5,
      avg_resolution_time_seconds: 144, // 2.4 minutes
      cost_savings_usd: 12300,
      total_tickets: 1250,
      auto_resolved: 980,
      escalated: 270,
    };
  }
  
  if (endpoint === "/agents") {
    return {
      agents: [
        {
          agent_name: "intent_detection",
          execution_count: 1250,
          successful_executions: 1180,
          failed_executions: 70,
          success_rate: 94.4,
          avg_latency_ms: 245,
          total_tokens: 12500,
          total_cost_usd: 0.42,
        },
        {
          agent_name: "diagnostic",
          execution_count: 1180,
          successful_executions: 1120,
          failed_executions: 60,
          success_rate: 94.9,
          avg_latency_ms: 320,
          total_tokens: 18500,
          total_cost_usd: 0.68,
        },
        {
          agent_name: "action_execution",
          execution_count: 1120,
          successful_executions: 1080,
          failed_executions: 40,
          success_rate: 96.4,
          avg_latency_ms: 180,
          total_tokens: 8200,
          total_cost_usd: 0.31,
        },
        {
          agent_name: "verification",
          execution_count: 1080,
          successful_executions: 1050,
          failed_executions: 30,
          success_rate: 97.2,
          avg_latency_ms: 150,
          total_tokens: 6500,
          total_cost_usd: 0.24,
        },
        {
          agent_name: "escalation_decision",
          execution_count: 1080,
          successful_executions: 1050,
          failed_executions: 30,
          success_rate: 97.2,
          avg_latency_ms: 120,
          total_tokens: 4200,
          total_cost_usd: 0.16,
        },
      ],
    };
  }
  
  if (endpoint.startsWith("/categories")) {
    return {
      categories: [
        {
          category_id: "printer_troubleshooting",
          category_name: "Printer Troubleshooting",
          date: new Date().toISOString().split("T")[0],
          total_tickets: 1250,
          successful_tickets: 1025,
          failed_tickets: 225,
          success_rate: 82.0,
          avg_latency_ms: 2100,
          total_tokens: 125000,
          total_cost_usd: 4.5,
        },
        {
          category_id: "hardware_diagnostics",
          category_name: "Hardware Diagnostics",
          date: new Date().toISOString().split("T")[0],
          total_tickets: 850,
          successful_tickets: 680,
          failed_tickets: 170,
          success_rate: 80.0,
          avg_latency_ms: 2800,
          total_tokens: 98000,
          total_cost_usd: 3.8,
        },
        {
          category_id: "network_connectivity",
          category_name: "Network Connectivity",
          date: new Date().toISOString().split("T")[0],
          total_tickets: 420,
          successful_tickets: 340,
          failed_tickets: 80,
          success_rate: 81.0,
          avg_latency_ms: 1900,
          total_tokens: 52000,
          total_cost_usd: 2.1,
        },
      ],
      resolution: {
        total: 2520,
        breakdown: {
          "auto-resolved": { count: 1960, percentage: 77.8, avg_mtr_seconds: 126 },
          escalated: { count: 480, percentage: 19.0, avg_mtr_seconds: 510 },
          failed: { count: 80, percentage: 3.2, avg_mtr_seconds: 0 },
        },
        // Category-specific breakdown
        by_category: {
          printer_troubleshooting: {
            total: 1250,
            auto_resolved: 975,
            escalated: 250,
            failed: 25,
            avg_mtr_auto: 126,
            avg_mtr_escalated: 480,
          },
          hardware_diagnostics: {
            total: 850,
            auto_resolved: 612,
            escalated: 204,
            failed: 34,
            avg_mtr_auto: 168,
            avg_mtr_escalated: 540,
          },
          network_connectivity: {
            total: 420,
            auto_resolved: 340,
            escalated: 68,
            failed: 12,
            avg_mtr_auto: 95,
            avg_mtr_escalated: 600,
          },
        },
      },
    };
  }
  
  if (endpoint === "/collaboration") {
    return {
      autonomous_agent_chain: {
        volume: 8450,
        success_rate: 94.0,
        avg_latency_ms: 1800,
        cost_per_resolution_usd: 0.12,
      },
      ai_to_human_handoff: {
        volume: 2280,
        handoff_rate: 21.3,
        avg_handoff_time_seconds: 3.2,
        human_resolution_rate: 88.0,
      },
      human_initiated_ai_assist: {
        volume: 1720,
        delegation_rate: 15.2,
        ai_success_rate: 89.0,
        time_saved_minutes: 4.2,
      },
    };
  }
  
  if (endpoint === "/insights") {
    return {
      insights: [
        {
          category_id: "printer_troubleshooting",
          category_name: "Printer Offline Issues",
          ai_handled: {
            resolution_rate: 88.0,
            avg_time_seconds: 168,
            accuracy: 91.0,
            cost_usd: 0.15,
          },
          human_handled: {
            resolution_rate: 82.0,
            avg_time_seconds: 432,
            csat: 4.3,
            cost_usd: 12.50,
          },
        },
        {
          category_id: "hardware_diagnostics",
          category_name: "Laptop Performance Issues",
          ai_handled: {
            resolution_rate: 82.0,
            avg_time_seconds: 216,
            accuracy: 86.0,
            cost_usd: 0.18,
          },
          human_handled: {
            resolution_rate: 76.0,
            avg_time_seconds: 570,
            csat: 4.1,
            cost_usd: 12.50,
          },
        },
        {
          category_id: "network_connectivity",
          category_name: "Network Connectivity Issues",
          ai_handled: {
            resolution_rate: 75.0,
            avg_time_seconds: 288,
            accuracy: 82.0,
            cost_usd: 0.20,
          },
          human_handled: {
            resolution_rate: 71.0,
            avg_time_seconds: 708,
            csat: 4.0,
            cost_usd: 12.50,
          },
        },
      ],
    };
  }
  
  if (endpoint === "/alerts") {
    return {
      alerts: [
        {
          type: "performance",
          severity: "warning",
          title: "High Latency Detected",
          description: "Diagnostic Agent averaging 3.2s (threshold: 2.0s) for Printer Offline workflows",
          timestamp: new Date(Date.now() - 8 * 60 * 1000).toISOString(),
        },
        {
          type: "volume_spike",
          severity: "info",
          title: "Laptop Issue Volume Spike",
          description: "Display flickering issues up 23% - investigating root cause",
          timestamp: new Date(Date.now() - 22 * 60 * 1000).toISOString(),
        },
        {
          type: "cost",
          severity: "error",
          title: "Token Usage Anomaly",
          description: "Knowledge Base Agent using 2.3x average tokens - possible prompt issue",
          timestamp: new Date(Date.now() - 35 * 60 * 1000).toISOString(),
        },
        {
          type: "success_rate",
          severity: "warning",
          title: "Success Rate Decline",
          description: "Network Issues category success rate dropped to 68% (7-day avg: 82%)",
          timestamp: new Date(Date.now() - 60 * 60 * 1000).toISOString(),
        },
      ],
    };
  }
  
  return {};
}

/**
 * Get entry channel volumes
 */
export async function getChannelMetrics(dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics("/channels", { from_date, to_date });
}

/**
 * Get key performance indicators
 */
export async function getKPIMetrics(dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics("/kpi", { from_date, to_date });
}

/**
 * Get agent performance data
 */
export async function getAgentMetrics(agentName = null, dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  const params = { from_date, to_date };
  if (agentName) {
    params.agent_name = agentName;
  }
  return fetchMetrics("/agents", params);
}

/**
 * Get agent execution details
 */
export async function getAgentDetails(agentName, dateRange = "7d", limit = 100) {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics(`/agents/${agentName}/details`, { from_date, to_date, limit });
}

/**
 * Get category performance data
 */
export async function getCategoryMetrics(categoryId = null, dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  const params = { from_date, to_date };
  if (categoryId) {
    params.category_id = categoryId;
  }
  return fetchMetrics("/categories", params);
}

/**
 * Get collaboration patterns
 */
export async function getCollaborationMetrics(dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics("/collaboration", { from_date, to_date });
}

/**
 * Get category insights
 */
export async function getInsights(dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics("/insights", { from_date, to_date });
}

/**
 * Get system alerts
 */
export async function getAlerts(dateRange = "7d") {
  const { from_date, to_date } = getDateRange(dateRange);
  return fetchMetrics("/alerts", { from_date, to_date });
}

File: backend/agentic_support/services/metrics_service.py
"""
Metrics service for aggregating and querying agent and category metrics
Author: Vinod Kumar V (VKV)
"""

import json
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Any

from ..db.database import get_db


class MetricsService:
    """Service for querying and aggregating metrics data."""
    
    @staticmethod
    def get_agent_metrics(
        agent_name: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """Get agent execution metrics."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = "SELECT * FROM agent_metrics WHERE 1=1"
            params = []
            
            if agent_name:
                query += " AND agent_name = ?"
                params.append(agent_name)
            
            if from_date:
                query += " AND timestamp >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND timestamp <= ?"
                params.append(to_date)
            
            query += " ORDER BY timestamp DESC LIMIT ?"
            params.append(limit)
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            return [dict(row) for row in rows]
    
    @staticmethod
    def get_agent_aggregates(
        agent_name: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get aggregated metrics for agents."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = """
                SELECT 
                    agent_name,
                    COUNT(*) as execution_count,
                    SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) as successful_executions,
                    SUM(CASE WHEN success = 0 THEN 1 ELSE 0 END) as failed_executions,
                    AVG(latency_ms) as avg_latency_ms,
                    SUM(tokens_total) as total_tokens,
                    SUM(cost_usd) as total_cost_usd
                FROM agent_metrics
                WHERE 1=1
            """
            params = []
            
            if agent_name:
                query += " AND agent_name = ?"
                params.append(agent_name)
            
            if from_date:
                query += " AND timestamp >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND timestamp <= ?"
                params.append(to_date)
            
            query += " GROUP BY agent_name"
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            results = []
            for row in rows:
                exec_count = row["execution_count"]
                success_count = row["successful_executions"]
                success_rate = (success_count / exec_count * 100) if exec_count > 0 else 0
                
                results.append({
                    "agent_name": row["agent_name"],
                    "execution_count": exec_count,
                    "successful_executions": success_count,
                    "failed_executions": row["failed_executions"],
                    "success_rate": round(success_rate, 2),
                    "avg_latency_ms": round(row["avg_latency_ms"] or 0, 2),
                    "total_tokens": row["total_tokens"] or 0,
                    "total_cost_usd": round(row["total_cost_usd"] or 0.0, 4),
                })
            
            return results
    
    @staticmethod
    def get_category_metrics(
        category_id: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Get category-level metrics."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = "SELECT * FROM category_metrics WHERE 1=1"
            params = []
            
            if category_id:
                query += " AND category_id = ?"
                params.append(category_id)
            
            if from_date:
                query += " AND date >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND date <= ?"
                params.append(to_date)
            
            query += " ORDER BY date DESC"
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            return [dict(row) for row in rows]
    
    @staticmethod
    def get_resolution_metrics(
        category_id: Optional[str] = None,
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get ticket resolution metrics with breakdown."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = """
                SELECT 
                    resolution_type,
                    COUNT(*) as count,
                    AVG(mtr_seconds) as avg_mtr_seconds
                FROM ticket_resolution_metrics
                WHERE 1=1
            """
            params = []
            
            if category_id:
                query += " AND category_id = ?"
                params.append(category_id)
            
            if from_date:
                query += " AND created_at >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND created_at <= ?"
                params.append(to_date)
            
            query += " GROUP BY resolution_type"
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            total = sum(row["count"] for row in rows)
            
            breakdown = {}
            for row in rows:
                resolution_type = row["resolution_type"] or "unknown"
                count = row["count"]
                breakdown[resolution_type] = {
                    "count": count,
                    "percentage": round((count / total * 100) if total > 0 else 0, 2),
                    "avg_mtr_seconds": round(row["avg_mtr_seconds"] or 0, 2),
                }
            
            return {
                "total": total,
                "breakdown": breakdown,
            }
    
    @staticmethod
    def get_handoff_metrics(
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get human handoff statistics."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = """
                SELECT 
                    handoff_type,
                    COUNT(*) as count
                FROM ticket_resolution_metrics
                WHERE handoff_type IS NOT NULL AND handoff_type != 'none'
            """
            params = []
            
            if from_date:
                query += " AND handoff_timestamp >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND handoff_timestamp <= ?"
                params.append(to_date)
            
            query += " GROUP BY handoff_type"
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            total_handoffs = sum(row["count"] for row in rows)
            
            # Get total tickets for percentage calculation
            total_query = "SELECT COUNT(*) as total FROM ticket_resolution_metrics WHERE 1=1"
            total_params = []
            if from_date:
                total_query += " AND created_at >= ?"
                total_params.append(from_date)
            if to_date:
                total_query += " AND created_at <= ?"
                total_params.append(to_date)
            
            cursor.execute(total_query, total_params)
            total_tickets = cursor.fetchone()["total"] or 0
            
            breakdown = {}
            for row in rows:
                handoff_type = row["handoff_type"]
                count = row["count"]
                breakdown[handoff_type] = {
                    "count": count,
                    "percentage": round((count / total_tickets * 100) if total_tickets > 0 else 0, 2),
                }
            
            return {
                "total_handoffs": total_handoffs,
                "handoff_percentage": round((total_handoffs / total_tickets * 100) if total_tickets > 0 else 0, 2),
                "breakdown": breakdown,
            }
    
    @staticmethod
    def get_channel_volumes(
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Get volume by entry channel."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            query = """
                SELECT 
                    channel,
                    COUNT(*) as volume,
                    SUM(CASE WHEN resolution_type = 'auto-resolved' THEN 1 ELSE 0 END) as auto_resolved,
                    SUM(CASE WHEN resolution_type = 'escalated' THEN 1 ELSE 0 END) as escalated
                FROM ticket_resolution_metrics
                WHERE 1=1
            """
            params = []
            
            if from_date:
                query += " AND created_at >= ?"
                params.append(from_date)
            
            if to_date:
                query += " AND created_at <= ?"
                params.append(to_date)
            
            query += " GROUP BY channel ORDER BY volume DESC"
            
            cursor.execute(query, params)
            rows = cursor.fetchall()
            
            return [
                {
                    "channel": row["channel"] or "unknown",
                    "volume": row["volume"],
                    "auto_resolved": row["auto_resolved"],
                    "escalated": row["escalated"],
                }
                for row in rows
            ]
    
    @staticmethod
    def get_kpi_metrics(
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get key performance indicators."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Self-heal success rate
            resolution_query = """
                SELECT 
                    COUNT(*) as total,
                    SUM(CASE WHEN resolution_type = 'auto-resolved' THEN 1 ELSE 0 END) as auto_resolved
                FROM ticket_resolution_metrics
                WHERE 1=1
            """
            resolution_params = []
            if from_date:
                resolution_query += " AND created_at >= ?"
                resolution_params.append(from_date)
            if to_date:
                resolution_query += " AND created_at <= ?"
                resolution_params.append(to_date)
            
            cursor.execute(resolution_query, resolution_params)
            resolution_row = cursor.fetchone()
            total_tickets = resolution_row["total"] or 0
            auto_resolved = resolution_row["auto_resolved"] or 0
            self_heal_rate = (auto_resolved / total_tickets * 100) if total_tickets > 0 else 0
            
            # Average resolution time
            mtr_query = """
                SELECT AVG(mtr_seconds) as avg_mtr
                FROM ticket_resolution_metrics
                WHERE mtr_seconds IS NOT NULL
            """
            mtr_params = []
            if from_date:
                mtr_query += " AND created_at >= ?"
                mtr_params.append(from_date)
            if to_date:
                mtr_query += " AND created_at <= ?"
                mtr_params.append(to_date)
            
            cursor.execute(mtr_query, mtr_params)
            mtr_row = cursor.fetchone()
            avg_mtr_seconds = mtr_row["avg_mtr"] or 0
            
            # Total cost savings (estimate: human support costs $12.50 per ticket, AI costs from agent_metrics)
            cost_query = """
                SELECT SUM(cost_usd) as total_cost
                FROM agent_metrics
                WHERE 1=1
            """
            cost_params = []
            if from_date:
                cost_query += " AND timestamp >= ?"
                cost_params.append(from_date)
            if to_date:
                cost_query += " AND timestamp <= ?"
                cost_params.append(to_date)
            
            cursor.execute(cost_query, cost_params)
            cost_row = cursor.fetchone()
            total_ai_cost = cost_row["total_cost"] or 0.0
            
            # Estimate human cost (escalated tickets * $12.50)
            escalated_count = total_tickets - auto_resolved
            estimated_human_cost = escalated_count * 12.50
            cost_savings = estimated_human_cost - total_ai_cost
            
            return {
                "self_heal_rate": round(self_heal_rate, 2),
                "avg_resolution_time_seconds": round(avg_mtr_seconds, 2),
                "cost_savings_usd": round(cost_savings, 2),
                "total_tickets": total_tickets,
                "auto_resolved": auto_resolved,
                "escalated": escalated_count,
            }
    
    @staticmethod
    def get_collaboration_metrics(
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get agent collaboration patterns."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Autonomous agent chain (tickets that were auto-resolved)
            auto_query = """
                SELECT COUNT(*) as count
                FROM ticket_resolution_metrics
                WHERE resolution_type = 'auto-resolved'
            """
            auto_params = []
            if from_date:
                auto_query += " AND created_at >= ?"
                auto_params.append(from_date)
            if to_date:
                auto_query += " AND created_at <= ?"
                auto_params.append(to_date)
            
            cursor.execute(auto_query, auto_params)
            auto_count = cursor.fetchone()["count"] or 0
            
            # Get success rate for auto-resolved (from agent metrics)
            agent_success_query = """
                SELECT 
                    COUNT(*) as total,
                    SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) as successful
                FROM agent_metrics
                WHERE 1=1
            """
            agent_params = []
            if from_date:
                agent_success_query += " AND timestamp >= ?"
                agent_params.append(from_date)
            if to_date:
                agent_success_query += " AND timestamp <= ?"
                agent_params.append(to_date)
            
            cursor.execute(agent_success_query, agent_params)
            agent_row = cursor.fetchone()
            agent_total = agent_row["total"] or 0
            agent_successful = agent_row["successful"] or 0
            agent_success_rate = (agent_successful / agent_total * 100) if agent_total > 0 else 0
            
            # Get avg latency and cost for auto-resolved
            agent_avg_query = """
                SELECT 
                    AVG(latency_ms) as avg_latency,
                    AVG(cost_usd) as avg_cost
                FROM agent_metrics
                WHERE success = 1
            """
            agent_avg_params = []
            if from_date:
                agent_avg_query += " AND timestamp >= ?"
                agent_avg_params.append(from_date)
            if to_date:
                agent_avg_query += " AND timestamp <= ?"
                agent_avg_params.append(to_date)
            
            cursor.execute(agent_avg_query, agent_avg_params)
            agent_avg_row = cursor.fetchone()
            
            # AI-to-Human handoff
            handoff_query = """
                SELECT 
                    COUNT(*) as count,
                    AVG(mtr_seconds) as avg_handoff_time
                FROM ticket_resolution_metrics
                WHERE resolution_type = 'escalated' AND handoff_type IS NOT NULL
            """
            handoff_params = []
            if from_date:
                handoff_query += " AND created_at >= ?"
                handoff_params.append(from_date)
            if to_date:
                handoff_query += " AND created_at <= ?"
                handoff_params.append(to_date)
            
            cursor.execute(handoff_query, handoff_params)
            handoff_row = cursor.fetchone()
            handoff_count = handoff_row["count"] or 0
            
            # Human resolution rate (assume 88% for escalated tickets)
            human_resolution_rate = 88.0
            
            return {
                "autonomous_agent_chain": {
                    "volume": auto_count,
                    "success_rate": round(agent_success_rate, 2),
                    "avg_latency_ms": round(agent_avg_row["avg_latency"] or 0, 2),
                    "cost_per_resolution_usd": round(agent_avg_row["avg_cost"] or 0.0, 4),
                },
                "ai_to_human_handoff": {
                    "volume": handoff_count,
                    "handoff_rate": round((handoff_count / (auto_count + handoff_count) * 100) if (auto_count + handoff_count) > 0 else 0, 2),
                    "avg_handoff_time_seconds": round(handoff_row["avg_handoff_time"] or 0, 2),
                    "human_resolution_rate": human_resolution_rate,
                },
                "human_initiated_ai_assist": {
                    "volume": 0,  # Placeholder - would need additional tracking
                    "delegation_rate": 0.0,
                    "ai_success_rate": 0.0,
                    "time_saved_minutes": 0.0,
                },
            }
    
    @staticmethod
    def get_alerts(
        from_date: Optional[str] = None,
        to_date: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Get system health alerts based on thresholds."""
        alerts = []
        
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Check for high latency agents
            latency_query = """
                SELECT agent_name, AVG(latency_ms) as avg_latency
                FROM agent_metrics
                WHERE timestamp >= datetime('now', '-1 hour')
                GROUP BY agent_name
                HAVING AVG(latency_ms) > 2000
            """
            cursor.execute(latency_query)
            latency_rows = cursor.fetchall()
            
            for row in latency_rows:
                alerts.append({
                    "type": "performance",
                    "severity": "warning",
                    "title": "High Latency Detected",
                    "description": f"{row['agent_name']} averaging {row['avg_latency']:.0f}ms (threshold: 2000ms)",
                    "timestamp": datetime.utcnow().isoformat(),
                })
            
            # Check for success rate decline
            success_query = """
                SELECT 
                    agent_name,
                    (SUM(CASE WHEN success = 1 THEN 1 ELSE 0 END) * 100.0 / COUNT(*)) as current_rate
                FROM agent_metrics
                WHERE timestamp >= datetime('now', '-24 hours')
                GROUP BY agent_name
                HAVING current_rate < 70
            """
            cursor.execute(success_query)
            success_rows = cursor.fetchall()
            
            for row in success_rows:
                alerts.append({
                    "type": "success_rate",
                    "severity": "error",
                    "title": "Success Rate Decline",
                    "description": f"{row['agent_name']} success rate dropped to {row['current_rate']:.1f}%",
                    "timestamp": datetime.utcnow().isoformat(),
                })
            
            # Check for volume spikes by category
            volume_query = """
                SELECT 
                    category_name,
                    COUNT(*) as recent_count
                FROM ticket_resolution_metrics
                WHERE created_at >= datetime('now', '-1 hour')
                GROUP BY category_name
                HAVING recent_count > 50
            """
            cursor.execute(volume_query)
            volume_rows = cursor.fetchall()
            
            for row in volume_rows:
                alerts.append({
                    "type": "volume_spike",
                    "severity": "info",
                    "title": "Volume Spike Detected",
                    "description": f"{row['category_name']} issues up - {row['recent_count']} tickets in last hour",
                    "timestamp": datetime.utcnow().isoformat(),
                })
        
        # Sort by timestamp (most recent first) and limit to 10
        alerts.sort(key=lambda x: x["timestamp"], reverse=True)
        return alerts[:10]
    
    @staticmethod
    def update_category_metrics(
        category_id: str,
        category_name: str,
        date: str,
    ):
        """Update or insert category metrics (called after ticket resolution)."""
        with get_db() as conn:
            cursor = conn.cursor()
            
            # Get ticket counts for this category and date
            resolution_query = """
                SELECT 
                    COUNT(*) as total,
                    SUM(CASE WHEN resolution_type = 'auto-resolved' THEN 1 ELSE 0 END) as successful,
                    SUM(CASE WHEN resolution_type = 'failed' THEN 1 ELSE 0 END) as failed,
                    AVG(mtr_seconds) as avg_mtr
                FROM ticket_resolution_metrics
                WHERE category_id = ? AND DATE(created_at) = ?
            """
            cursor.execute(resolution_query, (category_id, date))
            resolution_row = cursor.fetchone()
            
            # Get agent metrics for this category
            agent_query = """
                SELECT 
                    AVG(latency_ms) as avg_latency,
                    SUM(tokens_total) as total_tokens,
                    SUM(cost_usd) as total_cost
                FROM agent_metrics
                WHERE category_id = ? AND DATE(timestamp) = ?
            """
            cursor.execute(agent_query, (category_id, date))
            agent_row = cursor.fetchone()
            
            total = resolution_row["total"] or 0
            successful = resolution_row["successful"] or 0
            success_rate = (successful / total * 100) if total > 0 else 0
            
            # Upsert category metrics
            cursor.execute("""
                INSERT INTO category_metrics (
                    category_id, category_name, date,
                    total_tickets, successful_tickets, failed_tickets,
                    success_rate, avg_latency_ms, total_tokens, total_cost_usd, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(category_id, date) DO UPDATE SET
                    total_tickets = excluded.total_tickets,
                    successful_tickets = excluded.successful_tickets,
                    failed_tickets = excluded.failed_tickets,
                    success_rate = excluded.success_rate,
                    avg_latency_ms = excluded.avg_latency_ms,
                    total_tokens = excluded.total_tokens,
                    total_cost_usd = excluded.total_cost_usd,
                    updated_at = excluded.updated_at
            """, (
                category_id,
                category_name,
                date,
                total,
                successful,
                resolution_row["failed"] or 0,
                success_rate,
                int(agent_row["avg_latency"] or 0),
                agent_row["total_tokens"] or 0,
                agent_row["total_cost"] or 0.0,
                datetime.utcnow(),
            ))


# Global instance
metrics_service = MetricsService()
